python/revng/tupletree/__init__.py

```diff
@@ -152,8 +153,8 @@ def __post_init__(self):
                 setattr(self, field.name, new_field_value)
 
     def __setattr__(self, key, value):
-        # Prevent setting undefined attributes
-        if self.__dataclass_fields__.get(key) is None:
+        # Prevent setting undefined attributes, `_` allows to set private attributes like `_project`
+        if key not in [f.name for f in fields(self)] and not key.startswith("_"):
```

Comment:
Invert the conditions, e.g. `if not (key in [f.name for f in fields(self)] or key.startswith("_"))`